mixin-metrics.go

```diff
@@ -0,0 +1,319 @@
+	// slice of prom metrics in file
+	Metrics        []string	`json:"metrics"`
+	// promql parse errors in file
+	ParseErrors    []string `json:"parse_errors"`
+}
+
+// todo: dashboard structs https://github.com/grafana-tools/sdk/issues/130#issuecomment-797018658
+// using jq for now
+type RuleConfig struct {
+        RuleGroups	[]RuleGroup	`yaml:"groups"`
+}
+
+type RuleGroup struct {
+	Name     string     `yaml:"name"`
+	Rules    []Rule     `yaml:"rules"`
+}
+
+type Rule struct {
+	Record      string            `yaml:"record,omitempty"`
+	Alert       string            `yaml:"alert,omitempty"`
+	Expr        string            `yaml:"expr"`
+	Labels      map[string]string `yaml:"labels,omitempty"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
+}
+
+// create MetricsFile struct
+func NewMetricsFile(fn string, metrics map[string]struct{}, errs []error) MetricsFile {
+
+	keys := make([]string, 0, len(metrics))
+	for k := range metrics {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	errStrings := make([]string, 0, len(errs))
+	for _, err := range errs {
+		errStrings  = append(errStrings, err.Error())
+	}
+
+	metricsFile := MetricsFile{
+		Filename: fn,
+		Metrics: keys,
+		ParseErrors: errStrings,
+	}
+
+	return metricsFile
+}
+
+// use promql parser on a query to extract metrics
+func ParseQuery(query string, metrics map[string]struct{}) error {
+
+	query = strings.ReplaceAll(query, `\"`, `"`)
+	query = strings.ReplaceAll(query, `\n`, ``)
+	query = strings.ReplaceAll(query, `$__interval`, "5m")
+	query = strings.ReplaceAll(query, `$__rate_interval`, "5m")
+	query = strings.ReplaceAll(query, `$interval`, "5m")
+	query = strings.ReplaceAll(query, `$resolution`, "5s")
+
+	// label_values 
+	if strings.Contains(query, "label_values"){
+		re := regexp.MustCompile(`label_values\(([a-zA-Z0-9_]+)`)
+		query = re.FindStringSubmatch(query)[1]
+	}
+
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		err = errors.Wrapf(err, "promql query=%v", query)
+		log.Debugln("msg", "promql parse error", "err", err, "query", query)
+		return err
+	}
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		if n, ok := node.(*parser.VectorSelector); ok {
+			metrics[n.Name] = struct{}{}
+		}
+		return nil
+	})
+
+	return nil
+}
+
+// use jq on a file to extract prom queries
+func ParseJq(queries *[]string, jsonData map[string]interface{}, jqExpr string) error {
+
+	query, err := gojq.Parse(jqExpr)
+	if err != nil {
+		return err
+	}
+
+	iter := query.Run(jsonData)
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			break
+		}
+		if err, ok := v.(error); ok {
+			return err
+		}
+		*queries = append(*queries, v.(string))
+	}
+
+	return nil
+}
+
+// parses through a dashboard to extract queries, and then metrics from queries
+func ParseDashboard(file *os.File) (*MetricsFile, error) {
+
+	queries := make([]string, 0)
+	metrics := map[string]struct{}{}
+	errors := make([]error, 0)
+	res := map[string]interface{}{}
```

Comment:
Personal style nit: group variables defined together into a var block:

```suggestion
	var (
		queries = make([]string, 0)
		metrics = map[string]struct{}{}
		errors  = make([]error, 0)
		res     = map[string]interface{}{}
	)
```